<a href="https://colab.research.google.com/github/jagr0998/modelos_estadisticos/blob/main/EMV_e_Intervalos_de_Confianza.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Sala 7
* Guadarrama Ramírez José Angel
* Lobato Ceja María Elena
* Oros López Andrés
* Transito Crescencio Luis Daniel

# Estimador Máximo Verósimil

- Una v.a $X$, es exponencial con parámetro $\lambda >0$ si $$\mathbb{P}(X)=1-e^{-\lambda x}, x >0 $$
Calcular el EMV

- Supongamos que la vida útil de los focos de cierta marca se modela mediante una distribución exponencial con parámetro desconocido $\lambda$. Probamos 5 focos y encontramos que la vida útil de $2,3,1,3$ y $4$ años respectivamente. ¿Cual es el EMV para $\lambda$?
- Suponga que se tiene una variable aleatoria discreta $X$ con densidad:
	\begin{align*}
	P(X=0)&=\frac{2}{3}\theta,\\
	P(X=1)&=\frac{1}{3}\theta,\\
	P(X=2)&=\frac{2}{3}(1-\theta),\\
	P(X=3)&=\frac{1}{3}(1-\theta)
	\end{align*}
  con $\theta\in [0,1]$. Se toman las siguientes $10$ observaciones independientes de acuerdo a la distribución dada:
  $$(3, 0, 2, 1, 3, 2, 1, 0, 2, 1).$$
  Encuentra el EMV de $\theta$.

## Soluciones EMV



## 1.

Estimaremos λ
tenemos que $f(x,λ) = λe^{-λx}$

La función de verosimilitud esta dada por

$$ ⇒ L(λ) =  ∏_{i=1}^{n}f(x_{i},λ) = ∏_{i=1}^{n} λe^{-λx_{i}} = λ^{n}e^{-λ ∑_{i}x_{i}}$$

Aplicando $log$
$$ ⇒  log(λ^{n}e^{-λ ∑_{i}x_{i}}) = n log(λ) - λlog(e^{ ∑_{i}x_{i}})  = n log(λ) - λ∑_{i}x_{i} $$


$$⇒ \frac{d log(L(λ))}{dλ} = (n log(λ) - λ∑_{i}x_{i})' = 0$$
$$⇒(\frac{n}{λ} - ∑_{i}x_{i}) = 0$$
$$⇒ \hat{λ}_{EMV} = \frac{n}{∑_{i}x_{i}}  $$


## 2.
Por lo tanto en el ejemplo tenemos que,

$$⇒ \hat{λ}_{EMV} = \frac{n}{∑_{i}x_{i}} = \frac{5}{2+3+1+3+4} = \frac{5}{13}$$

In [1]:
x = [2,3,1,3,4]
n = 5
suma = sum (x)

lamhat = n/suma
print(lamhat)

0.38461538461538464


## 3.
Tenemos la muestra (3, 0, 2, 1, 3, 2, 1, 0, 2, 1)

$$ ⇒ L(θ) =  ∏_{i=1}^{n}f(x_{i},θ) = f(0,θ)^{2}f(1,θ)^{3}f(2,θ)^{3}f(3,θ)^{2}$$

$$ ⇒ L(θ) = (\frac{2}{3}θ)^{2}(\frac{1}{3}θ)^{3}(\frac{2}{3}(1-θ))^{3}(\frac{1}{3}(1-θ))^{2} = \frac{2^{2+3}}{3^{2+3+3+2}}(θ)^{2+3}(1-θ)^{2+3} = \frac{2^{5}}{3^{10}}(θ)^{5}(1-θ)^{5}$$

Aplicando $log$
$$ ⇒  log(\frac{2^{5}}{3^{10}}(θ)^{5}(1-θ)^{5}) =  log(\frac{2^{5}}{3^{10}}) + 5log(θ) + 5log(1-θ)$$


$$⇒ \frac{d log(L(θ))}{dθ} = (log(\frac{2^{5}}{3^{10}}) + 5log(θ) + 5log(1-θ))' = 0$$

$$⇒\frac{5}{θ} - \frac{5}{1-θ} = 0$$

$$⇒\frac{5}{θ} = \frac{5}{1-θ} $$

$$⇒5-5θ = 5θ $$



Por lo tanto en el ejemplo tenemos que,

$$⇒ \hat{θ}_{EMV} = \frac{1}{2} $$

In [2]:
from scipy.stats import expon
import numpy as np
import sympy # para calculo simbolico y algebraico
from sympy.abc import x,z,n

# Vamos a crear una variable simbolica p
t = sympy.symbols('t',positive=True)

f0 = (2/3)*t
f1 = (1/3)*t
f2 = (2/3)*(1-t)
f3 = (1/3)*(1-t)
J = (2*f0)*(3*f1)*(3*f2)*(2*f3) #funcion de verosimilitud
logJ = sympy.expand_log(sympy.log(J)) #log de la verosimilitud
that = sympy.solve(sympy.diff(logJ,t),t)[0] #EMV
print("El estimador de maxima verosimilitud es", that)

El estimador de maxima verosimilitud es 0.500000000000000


# Intervalos de confianza

- En una encuesta de la UP, 225 estudiantes son seleccionados al azar de los que toman cálculo y se les pregunta si cálculo es su materia favorita. 100 estudiantes responden que el cálculo es su materia favorita. Proporcione un intervalo de confianza del 95 %  para la proporción de todos los estudiantes que toman cálculo y que la consideran su materia favorita.  

- Los datos
	$$1, 2, 3, 4$$
  se toman a partir de una distribución normal $N(\mu,\sigma^2)$ con $\mu$ desconocida.
  1. Encuentra un intervalo de confianza al nivel de $90\%$ para $\mu$, dado que $\sigma =2$.
  2. Encuentra un intervalo de confianza al nivel de $90\%$ para $\mu$.
  3. Encuentra un intervalo de confianza al nivel de $90\%$ para $\sigma^2$.

In [6]:
import scipy.stats as stats
import statsmodels.api as sm
import numpy as np
import scipy.stats as st

k=4
#1
alpha = 1-.95
z_1 = round(stats.norm.ppf(q = 1-alpha/2), 2)

#2.1
alpha = 1-.9
z_2 = round(stats.norm.ppf(q = 1-alpha/2), 2)

#2.2
t = round(stats.t.ppf(1-alpha/2, k-1), 2)

#2.3
x_1 = round(stats.chi2.ppf(1-alpha/2, k-1), 2)
x_2 = round(stats.chi2.ppf(alpha/2, k-1), 2)

In [ ]:

print(z_1, z_2, t, x_1, x_2)

1.96 1.64 2.35 7.81 0.35


## Soluciones IC

## 1.Proporcione un intervalo de confianza del 95 %  para la proporción de todos los estudiantes que toman cálculo y que la consideran su materia favorita.

Tenemos ciertos datos, como el tamaño de la muestra $n = 225 $ y $x = 100$ (estudiantes que toman cálculo y que la consideran su materia favorita), asi que podemos tomar la proporción $\hat{p}$ como:
$$\hat{p} = \frac{x}{n} ≈ .44$$

Tambien sabemos que la desviación estandár de la proporción esta dada por
$$ σ_{\hat{p}} = \sqrt{\frac{\hat{p}(1-\hat{p})}{n}}$$

$$ σ_{\hat{p}} = \sqrt{\frac{\frac{100}{225}(\frac{125}{225})}{225}}≈ .03$$

El valor critico $z$ para un nivel de confianza del 95%
$$ z_{α/2} = z_{.05/2} ≈ 1.96$$

Nuestro intervalo de confianza esta dado por


$$ IC = (\hat{p}-zσ_{\hat{p}}, \hat{p}+zσ_{\hat{p}})$$

$$ IC = (\frac{100}{225}-(1.96)\sqrt{\frac{\frac{100}{225}(\frac{125}{225})}{225}}, \frac{100}{225}+(1.96)\sqrt{\frac{\frac{100}{225}(\frac{125}{225})}{225}})$$

Por lo tanto, el IC al 95% de confianza es aproximadamente $(0.379, 0.509)$ para la proporción de todos los estudiantes que toman cálculo y que la consideran su materia favorita


In [5]:
p = 100
confidence = 0.95
n = 225

conf_int = sm.stats.proportion_confint(p, n, alpha=(1-confidence), method='normal')

print("Intervalo de confianza:", conf_int)

Intervalo de confianza: (0.37951684884617914, 0.5093720400427098)


## 2.1 Estimación de la media con $σ$ conocida


Tenemos ciertos datos, como el tamaño de la muestra $n = 4 $ y que distribución normal $N(μ,σ^2)$ dado que $σ =2$, asi que podemos calcular la media muestral


$$\bar{x} = \frac{1+2+3+4}{n} = \frac{10}{4} = 2.5$$

El valor critico $z$ para un nivel de confianza del 90%
$$ z_{α/2} = z_{.10/2} ≈ 1.64$$

Nuestro intervalo de confianza esta dado por
$$ IC = (\bar{x}-z\frac{σ}{\sqrt{n}}, \bar{x}+z\frac{σ}{\sqrt{n}})$$

$$ IC = (2.5-(1.64)\frac{2}{\sqrt{4}}, 2.5+(1.64)\frac{2}{\sqrt{4}})$$

Por lo tanto, el IC al 90% de confianza es aproximadamente $(0.86, 4.14)$ para la media con σ conocida


In [7]:
data = [1,2,3,4]
# Tamaño de la muestra
n = len(data)

# Media muestral
mean = np.mean(data)

# Desviación estándar conocida (sigma)
sigma = 2
confidence = 0.90
# Nivel de confianza
alpha = 1-confidence
z = st.norm.ppf(1 - alpha / 2)  # Valor crítico de z

# Calcular el intervalo de confianza
margin_of_error = z * (sigma / np.sqrt(n))
conf_int = (mean - margin_of_error, mean + margin_of_error)

print("Intervalo de confianza para la media:", conf_int)

Intervalo de confianza para la media: (0.8551463730485278, 4.144853626951472)



## 2.2 Estimación de la media con $σ$ desconocida

Sabemos que el tamaño de la muestra $n = 4 $ y la media muestral es $\bar{x}=2.5 $

Calculemos $s$ de la muestra

$$ ⇒ s = \sqrt{\frac{(\bar{x}-1)^{2} + (\bar{x}-2)^{2} + (\bar{x}-3)^{2}+ (\bar{x}-4)^{2}}{n-1}}$$

$$ ⇒ s = \sqrt{\frac{(2.5-1)^{2} + (2.5-2)^{2} + (2.5-3)^{2}+ (2.5-4)^{2}}{3}}$$

$$ ⇒ s ≈ 1.291$$


El valor critico $t$ para un nivel de confianza del 90%
$$ t_{\frac{α}{2}, n-1} = t_{.05, 3} ≈ 2.35$$

Nuestro intervalo de confianza esta dado por
$$ IC = (\bar{x}-t\frac{s}{\sqrt{n}}, \bar{x}+t\frac{s}{\sqrt{n}})$$

$$ IC = (2.5-(2.35)\frac{1.291}{\sqrt{4}}, 2.5+(2.35)\frac{1.291}{\sqrt{4}})$$

Por lo tanto, el IC al 90% de confianza es aproximadamente $(0.983, 4.016)$ para la media con σ desconocida



In [8]:
data = [1,2,3,4]
confidence = 0.90
gl = len(data) - 1
mean = np.mean(data)
error_est = st.sem(data)

# Intervalo para la media

IC = st.t.interval(confidence,gl,loc=mean,scale=error_est)
print("El intervalo del 90% de confianza para el peso medio de la poblacion es:", IC)

El intervalo del 90% de confianza para el peso medio de la poblacion es: (0.9809104349065063, 4.019089565093493)


## 2.3 Estimación de $σ^{2}$

Sabemos que el tamaño de la muestra $n = 4 $ y la media muestral es $\bar{x}=2.5 $

Calculemos $s^{2}$ de la muestra

$$ ⇒ s^{2} = \frac{(\bar{x}-1)^{2} + (\bar{x}-2)^{2} + (\bar{x}-3)^{2}+ (\bar{x}-4)^{2}}{n-1}$$

$$ ⇒ s^{2} = \frac{2.25+.25+.25+2.25}{3}$$

$$ ⇒ s^{2} = \frac{5}{3}$$


Los valores criticos de $χ^{2}$ para un nivel de confianza del 90% son
$$ χ^{2}_{\frac{α}{2}, n-1} = χ^{2}_{.05, 3} ≈ 7.81$$

$$ χ^{2}_{1-\frac{α}{2}, n-1} = χ^{2}_{.95, 3} ≈ .35$$

Nuestro intervalo de confianza esta dado por
$$ IC = (\frac{(n-1)s^{2}}{χ^{2}_{\frac{α}{2}, n-1}}, \frac{(n-1)s^{2}}{χ^{2}_{1-\frac{α}{2}, n-1}})$$

$$ IC = (\frac{(3)\frac{5}{3}}{7.81}, \frac{(3)\frac{5}{3}}{.35})$$

Por lo tanto, el IC al 90% de confianza es aproximadamente $(0.64, 14.28)$ para $σ^{2}$

In [9]:
# Varianza muestral
s2 = np.var(data, ddof=1)  # ddof=1 para la varianza muestral

# Nivel de confianza
alpha = 1-confidence

# Valores críticos de chi-cuadrado
chi2_lower = st.chi2.ppf(1 - alpha / 2, n - 1)
chi2_upper = st.chi2.ppf(alpha / 2, n - 1)

# Calcular el intervalo de confianza para la varianza
conf_int = ( (n - 1) * s2 / chi2_lower, (n - 1) * s2 / chi2_upper)

print("Intervalo de confianza para sigma^2:", conf_int)

Intervalo de confianza para sigma^2: (0.6398175421974498, 14.210749829597598)
